`init` accessors.

hborla opened the pull request:
Pitched here: https://forums.swift.org/t/pitch-init-accessors/64881

DougGregor commented on proposals/NNNN-init-accessors.md:
```diff
@@ -57,7 +57,7 @@ This proposal adds _`init` accessors_ to opt computed properties on types into v
   var degrees: Double
   var radians: Double {
-    init(newValue, initializes: degrees) {
+    init(newValue) initializes(degrees) {
```
I think we should call this `initialValue` rather than `newValue`.

Jumhyn commented on proposals/NNNN-init-accessors.md:
````diff
@@ -0,0 +1,362 @@
+}
+```
+
+A memberwise initializer cannot be synthesized if a stored property that is an `accesses` effect of a computed property is ordered after that computed property in the source code:
````
I left a comment to this effect in the pitch but I think 'cannot' is too strong here. I think topologically ordering the properties for initialization in the memberwise `init` body is perfectly sound. Open to being convinced that I'm wrong about soundness, or that this behavior is somehow confusing and undesirable, but think 'will not' is better verbiage here. I can save further comment for the review thread ;)

Jumhyn commented on proposals/NNNN-init-accessors.md:
````diff
@@ -0,0 +1,362 @@
+    self.point = (x, y) // error: neither the `init` accessor nor the setter can be called here
+  }
+}
+```
````
A case that I don't think is covered here: what about a computed property with an `init` accessor which is _conditionally_ initialized along some path, and then reassigned?

```
struct S {
    var x: Int {
        init initializes(y) {
            y = newValue
        }
        get { ... }
        set { ... }
    }
    var y: Int

    init(x: Int, y: Int) {
        if y > 0 {
            self.x = 0
        }

        self.x = x
    }
}

hborla replied:
An assignment to a property that is not initialized on all paths will call the `init` accessor again. This is a generalization of what property wrappers do. I can add this example to the proposal.

Jumhyn commented on proposals/NNNN-init-accessors.md:
```diff
@@ -0,0 +1,362 @@
+
+### Memberwise initializers
+
+If a struct does not declare its own initializers, it receives an implicit memberwise initializer based on the stored properties of the struct, because the storage is what needs to be initialized. Because `init` provide a preferred mechanism for initializing storage, the memberwise initializer parameter list will include any computed properties that subsume the initialization of stored properties ...
```
> Because `init` provide a preferred mechanism for initializing storage, the memberwise initializer parameter list will include any computed properties that subsume the initialization of stored properties instead of parameters for those stored properties.

I think this is a perfectly reasonable position, but I wonder if the notion that `init` accessors represent the preferred initialization strategy deserves mention as part of **Proposed solution** rather than just being a side mention in a subsection of **Detailed design**. That's a not-insignificant policy decision about the feature!

Jumhyn replied:
Concretely, I think the "preferred mechanism" thing (and its relationship to the memberwise init) is the only thing in this proposal that casts doubt on what otherwise seems a perfectly reasonable use of `init` accessors: having _multiple_ computed properties which initialize the same underlying storage, e.g., `radians` and `rotations` both initializing `degrees`.

hborla replied:
I added more commentary in the Proposed Solution and in this section about member-wise initializers.

Jumhyn commented on proposals/NNNN-init-accessors.md:
````diff
@@ -0,0 +1,362 @@
+}
+```
+
+Note that macro-expanded declarations are ordered after the attached-to declaration for peer macros, and at the end of the member list as written for member macros.
````
This line feels kind of like a non-sequitur and would probably benefit from an example. I'm not quite sure what it's saying or why it's germane in the discussion of `init` accessors.

hborla commented on proposals/NNNN-init-accessors.md:
```diff
@@ -0,0 +1,362 @@
+* Authors: [Holly Borla](https://github.com/hborla), [Doug Gregor](https://github.com/douggregor)
+* Review Manager: TBD
+* Status: **Awaiting implementation**
+
```
Yes, thank you!

Jumhyn commented on proposals/NNNN-init-accessors.md:
```diff
@@ -0,0 +1,363 @@
+* Authors: [Holly Borla](https://github.com/hborla), [Doug Gregor](https://github.com/douggregor)
+* Review Manager: TBD
+* Status: **Awaiting implementation**
+Implementation: On `main` behind experimental feature flag `InitAccessors`
```
```suggestion
* Implementation: On `main` behind experimental feature flag `InitAccessors`
```
oops, my bad 🤦

hborla commented on proposals/NNNN-init-accessors.md:
```diff
@@ -0,0 +1,362 @@
+
+## Proposed solution
+
+This proposal adds _`init` accessors_ to opt computed properties on types into virtual initialization that subsumes initialization of a set of zero or more specified stored properties, which allows assigning to computed properties in the body of a type's initializer:
```
>Is 'virtual' meant to be precise about the fact that the initialization is not (necessarily) tied to memory initialization of some storage location?

Yes, exactly. But I agree it's probably an unnecessary bit of jargon, and I can replace the phrase with something more direct.

Jumhyn replied:
Yeah I think that's perhaps an interesting aside to include somewhere in this document but IMO we don't need to harp on the fact that this is 'virtual' initialization since part of the _point_ of this proposal is to lift DI into a language concept _separate_ from just the stored properties that need to be initialized for memory safety reasons. I think we should instead orient towards the fact that computed properties with `init` accessors are, for DI purposes, treated the _same_ as stored properties rather than using language to set them apart throughout.

hborla replied:
I removed the phrase "virtual initialization" throughout the document.

hborla commented on proposals/NNNN-init-accessors.md:
```diff
@@ -0,0 +1,362 @@
+  }
+
+  init(radians: Double) {
+    self.radians = radians // calls init accessor with 'radians'
```
Heh, yeah I meant to describe that `radians` is passed as the argument to the init accessor, but I can see how that's completely unclear. I can change the name of the parameter so that it's more obvious.

hborla commented on proposals/NNNN-init-accessors.md:
```diff
@@ -0,0 +1,362 @@
+
+access-effect -> 'accesses' '(' identifier-list ')'
+
+identifier-list -> identifier
+identifier-list -> identifier ',' identifier-list
```
It might not be, but I didn't find an existing one while doing a completely-non-comprehensive scan over the existing syntax nodes in SwiftSyntax 🙂

Jumhyn replied:
It looks like we already have an entry for this production in [TSPL](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/lexicalstructure#Identifiers) so can probably omit from the proposal text!

> _identifier-list_ → _identifier_ | _identifier_ , _identifier-list_

hborla commented on proposals/NNNN-init-accessors.md:
````diff
@@ -0,0 +1,362 @@
+
+Access effects allow a computed property to be initialized by placing its contents into another stored property:
+
+```swift
````
That's a great point, I'll change the example to remove the initial value for `dictionary`!

hborla commented on proposals/NNNN-init-accessors.md:
````diff
@@ -0,0 +1,362 @@
+    set { _x = newValue }
+  }
+}
+```
````
Added.

hborla commented on proposals/NNNN-init-accessors.md:
```diff
@@ -0,0 +1,362 @@
+
+### `init` accessor signatures
+
+`init` accessor declarations can optionally specify a signature. An `init` accessor signature is composed of a parameter for the initial value, a list of stored properties that are initialized by this accessor specified with the contextual `initializes` keyword, and a list of stored properties that are accessed by this accessor specified with the contextual `accesses` keyword, all of which are op ...
```
Reworded.

hborla commented:
@Jumhyn I believe I've addressed all of your feedback, and I strengthened the Alternatives Considered section with some suggestions from the pitch thread. Please let me know if you have any further feedback!
